Add Ant build support for Schematron

cirulls opened the pull request:
## Summary

This pull request integrates #138 submitted by @markdunnoup and allows to run XSpec tests for Schematron using the ant build script. For further detail, please see conversation in #138.

This pull request also adds two bats tests checking that the ant build script works correctly for XSLT and Schematron. 

As this was approved by @markdunnoup and me in https://github.com/xspec/xspec/pull/138#issuecomment-360114612, I'm merging into master. 

## Further work

I will update the [documentation in the wiki](https://github.com/xspec/xspec/wiki/Running-with-Ant).

@AirQuick: would you be able to add similar tests for the batch script? Note that ant is installed by default in Travis but I had to install additional ant libraries in `.travis.yaml`. Ant may [not be installed by default on AppVeyor](http://help.appveyor.com/discussions/questions/1956-is-ant-part-of-the-default-oss-image).

AirQuick commented on build.xml:
```diff
@@ -8,119 +8,268 @@
+          name="http://saxon.sf.net/feature/allow-external-functions" 
+          value="true"/>
+      </factory>
+    </xslt>
```
`saxon.custom.options` is defined before, but not used here. Why? (Is it possible in the first place?)

cirulls replied:
I guess `saxon.custom.options` is used by oXygen. I can see it is defined with this comment:

```xml
  <!-- OXYGEN PATCH needed to create a custom Saxon, for example  -warnings:recover -strip:none -opt:10 -dtd:off -l:off -versionmsg:off -expand:on -outval:fatal -val:lax -->
  <property name="saxon.custom.options" value=""/>
  <!-- OXYGEN PATCH END -->
```

I wonder if it is is worth removing oXygen customizations that are currently not needed for this feature and work with oXygen for a smooth integration of XSpec. This should also address https://github.com/xspec/xspec/pull/138#issuecomment-360111771 raised by @markdunnoup.

Shall we just remove the `saxon.custom.options` if not used? Do you see any side effect?

AirQuick replied:
We may just remove it, but...
This PR changes `<java>` to `<xslt>`. Is it easy to implement the feature of `saxon.custom.options` in `<xslt>`? I'm afraid not.

cirulls replied:
I reintroduced the use of `<java>` instead of `<xslt>` in https://github.com/xspec/xspec/commit/3bcb2cd3ecaa6ff117e3193546673941544eff4f#diff-2cccd7bf48b7a9cc113ff564acd802a8R227. I have kept `saxon.custom.options` as a result.

AirQuick commented on build.xml:
```diff
@@ -8,119 +8,268 @@
+      force="true"
+      >     
+      <factory name="net.sf.saxon.TransformerFactoryImpl" />     
+    </xslt>
```
oXygen's `build.xml` enables XML catalog here and in other places. Whey are they not incorporated in this PR?

cirulls replied:
I reintroduced XML catalog in various places in https://github.com/xspec/xspec/commit/3bcb2cd3ecaa6ff117e3193546673941544eff4f following the current version of `build.xml` packaged with oXygen 19. I had to add `xml-resolver.jar` as additional jar library in `.travis.yaml` and `appveyor.yaml` otherwise tests failed on my local machine with the following error message:

 ```
[java] Transformation failed: Failed to load Apache catalog resolver library [java]
```

Tests are passing on [Travis](https://travis-ci.org/xspec/xspec/builds/333452222) but failing on [AppVeyor](https://ci.appveyor.com/project/xspec/xspec/build/1.0.334/job/db55wg40d1rhayof). The culprit is the ant test for Schematron which fails with this error:

```
[java] Error on line 12 column 87 of demo-03.xsl:
     [java]   Malformed URL C:/projects/xspec/tutorial/schematron/demo-03-sch-compiled.xsl(base
     [java]   file:/C:/projects/xspec/tutorial/schematron/xspec/demo-03.xsl): unknown protocol: c
     [java] Malformed URL C:/projects/xspec/tutorial/schematron/demo-03-sch-compiled.xsl(base file:/C:/projects/xspec/tutorial/schematron/xspec/demo-03.xsl)
``` 

The issue is the path `file:/C`, I suspect the XML resolver or the java command may have caused this. I also wonder if this is a similar problem I described in https://github.com/xspec/xspec/pull/138#pullrequestreview-53085196 (see where I talk about `file:/` in Unix and Windows) but this time it's problematic for Windows. 

I won't have more time today to debug this and unfortunately Windows is not my forte. Could you have a look at this failing test in case you spot any glaring error I made?

AirQuick replied:
Thanks for the improvements!
I looked at the intermediate file (`demo-03.xsl`) and found a fake URI:
```xml
   <xsl:import href="c:/xspec/tutorial/schematron/demo-03-sch-compiled.xsl"/>
...
   <xsl:variable name="x:stylesheet-uri"
                 as="xs:string"
                 select="'c:/xspec/tutorial/schematron/demo-03-sch-compiled.xsl'"/>
```
Probably https://github.com/xspec/xspec/pull/160/commits/4e2ed51f98b7f530b21bbbc3742eb97fe9712294#diff-2cccd7bf48b7a9cc113ff564acd802a8R129 makes it.
I'm fixing it.

AirQuick commented on build.xml:
```diff
@@ -8,119 +8,268 @@
+    xspec.phase                Required if XSpec describes a Schematron file, with value "#ALL" 
+    xspec.compiled.xsl.dir     Folder in which to place the compiled XSLT file generated from a Schematron file
+                               [Optional; the default is a subfolder "xspec" of the folder containing the XSpec file]
+    
```
Not all parameters are explained. Are they missed? Or, will be explained only in Wiki?

cirulls replied:
I added more documentation in https://github.com/xspec/xspec/commit/e9861f8045f20806b05725e3b0cf53ea1180732e. This will be used to update the wiki. I left the documentation on how to run it in oXygen as it will be useful for the oXygen integration and for all those who want to use it before this feature is fully integrated in oXygen.

AirQuick commented on build.xml:
```diff
@@ -8,119 +8,268 @@
+  
+  <!-- Copy of original XSpec file
+  will be overwritten if XSpec applies to Schematron -->
+  <property name="xspec.compiled.xml" location="${xspec.base.dir}/${xspec.base}-compiled-oxygen.xspec"/>
```
Why does this file name have `-oxygen`?
It was puzzling to find such a file when I ran `build.xml` outside oXygen.

cirulls replied:
I removed the `-oxygen` from the file name in https://github.com/xspec/xspec/commit/3b1a6f49032aef61941c06a5fd5bb6eeb82724d6#diff-2cccd7bf48b7a9cc113ff564acd802a8R113. This should be unnecessary once we fully integrate with  the oXygen ant build file.

AirQuick commented on test/xspec.bats:
```diff
@@ -326,3 +326,20 @@ teardown() {
+
+
+@test "running XSpec via ant for XSLT support" {
+    run ant -buildfile ${PWD}/../build.xml -Dxspec.xml=${PWD}/../tutorial/escape-for-regex.xspec -lib /tmp/xspec/saxon/saxon9he.jar -lib /tmp/ant/lib/xmltask.jar -lib /tmp/ant/lib/ant-contrib.jar 
```
* `-lib /tmp/xspec/saxon/saxon9he.jar` should be `-lib ${SAXON_CP}` to coordinate with other tests and help changing Saxon versions/editions.
* `-lib /tmp/ant/lib/xmltask.jar -lib /tmp/ant/lib/ant-contrib.jar` may be shortened to `-lib /tmp/ant/lib`
* `-Dclean.output.dir=true` should be added, otherwise `git status` finds a temp file after testing locally.

cirulls replied:
I implemented this feedback in https://github.com/xspec/xspec/commit/cd327716028c1071ec2f4a89c4cc072fb98e65a0

AirQuick commented on test/xspec.bats:
```diff
@@ -326,3 +326,20 @@ teardown() {
+
+@test "running XSpec via ant for Schematron support" {
+    echo ${PWD}
+    run ant -buildfile ${PWD}/../build.xml -Dxspec.xml=${PWD}/../tutorial/schematron/demo-03.xspec -lib /tmp/xspec/saxon/saxon9he.jar -lib /tmp/ant/lib/xmltask.jar -lib /tmp/ant/lib/ant-contrib.jar -Dtest.type=s -Dxspec.project.dir=${PWD}/../ -Dxspec.compiled.xsl.dir=${PWD}/../tutorial/schematron -Dxspec.phase=#ALL
```
After running this command line locally, `git status` finds `tutorial/schematron/demo-03-sch-compiled.xsl` even with `-Dclean.output.dir=true`. Is it a bug of `clean.output.dir`?

cirulls replied:
I remember @markdunnoup mentioning this in https://github.com/xspec/xspec/pull/138#issue-236885420:

>The "cleanup" target is not fully compatible with XSpec for Schematron yet. When ${clean.output.dir} is set to true, it removes SVRL files which the HTML report links to if the XSPec fails.

AirQuick replied:
The problem here is that a file is *not* removed. The known problem in the cited comment is that a file *is* removed. Are they the same thing?
Anyway, if `cleanup` is incomplete, it would be nice if known issues or remaining tasks are commented in the file.

cirulls replied:
I added the removal of the temporary file in https://github.com/xspec/xspec/commit/3b1a6f49032aef61941c06a5fd5bb6eeb82724d6#diff-2cccd7bf48b7a9cc113ff564acd802a8R252, the XSL compiled file was missing.

AirQuick commented:
> @AirQuick: would you be able to add similar tests for the batch script?

I pushed the tests to this branch.

* `appveyor.yml` on AppVeyor defines an environment variable [`XSPEC_ANT_LIB`](https://github.com/xspec/xspec/pull/160/commits/7243d0c225fa54dd0f2b05e749a7142dc5012851#diff-180360612c6b8c4ed830919bbb4dd459R44) and installs external Jar files there.
* `xspec-bat.cmd` runs Ant tests only when the `XSPEC_ANT_LIB` is defined. ([XSLT](https://github.com/xspec/xspec/pull/160/commits/45a005f182e81c79afe71cf7883fd63eb5ee85b3#diff-78e10853d6dbb71bb7fcc723c9791f94R397) and [Schematron](https://github.com/xspec/xspec/pull/160/commits/45a005f182e81c79afe71cf7883fd63eb5ee85b3#diff-78e10853d6dbb71bb7fcc723c9791f94R411))

I made so, because I usually run `xspec-bat.cmd` on a local computer where Ant and libraries are not readily available.

AirQuick commented on test/xspec.bats:
```diff
@@ -326,3 +326,20 @@ teardown() {
+
+
+@test "running XSpec via ant for XSLT support" {
+    run ant -buildfile ${PWD}/../build.xml -Dxspec.xml=${PWD}/../tutorial/escape-for-regex.xspec -lib ${SAXON_CP} -lib /tmp/ant/lib 
```
If I recall correctly, `-Dclean.output.dir=true` is necessary here too.

cirulls replied:
Good catch! I added it in https://github.com/xspec/xspec/commit/ab86e9fb7dcc1e4c56b6f12bab57447e586df5bb

cirulls commented on build.xml:
```diff
@@ -8,119 +8,268 @@
-			</antcall>
-		</sequential>
-	</macrodef>
+  <!-- OXYGEN PATCH needed to create a custom Saxon, for example  -warnings:recover -strip:none -opt:10 -dtd:off -l:off -versionmsg:off -expand:on -outval:fatal -val:lax -->
```
I removed the `OXYGEN PATCH` title but kept the comment for documentation purposes in https://github.com/xspec/xspec/commit/3b1a6f49032aef61941c06a5fd5bb6eeb82724d6

AirQuick commented:
After fixing https://github.com/xspec/xspec/pull/160#discussion_r164039512, **ant-contrib** seems no longer necessary. So I removed it.

AirQuick commented:
**xmltask** is not necessary for testing XSLT. Always requiring its jar is a bit annoying. So I made it optional.

AirQuick commented:
`<fail>` was removed by a patch from oXygen with a comment `<!-- OXYGEN PATCH the result will be opened and the user can inspect it. A fail is too much -->`.
It's true that a fail is too much when testing on oXygen. But what if this build file is used as a part of a real build process?
I'd think it better to keep `<fail>` and disables it conditionally by a system property.

AirQuick commented:
**xml-resolver** is necessary only for testing with catalog. So I made it optional.

AirQuick commented:
I found `catalog` does not always work. So I [fixed it](https://github.com/xspec/xspec/pull/160/commits/2f86d7dbd8ea7babacf6ffb40891932cd27161a0).

oXygen 19 default transformation scenario sets `catalog` property = `${pdu}/catalog.xml`. After integrating this change, it must be changed to `${pd}/catalog.xml`.

AirQuick commented:
`copyXSpec` leaves a file outside `xspec.dir` on error of XSLT being tested. A bit annoying.
So I took a look at it and I found `copyXSpec` unnecessary in the first place. So I removed it and added some conditions instead.

cirulls commented:
Thanks for your sterling work @AirQuick, you are making great improvements and simplifying a lot the ant build script. I checked the commit changes you made so far and they look good to me. Let me know when you're done and when you think it's ready to be merged and I'll have a final look.

AirQuick commented:
I'm done except for https://github.com/xspec/xspec/pull/160#issuecomment-360747550 regarding `<fail>`. What do you think?
(I'm going to bed now and will be dormant for half a day.)

cirulls commented:
I'll have a look tomorrow morning (UK time) and let you know. Have a well-deserved sleep!

AirQuick commented:
I encountered some problems in Schematron handler in this build file. I'll fix it. Fixing it will remove dependency on **xmltask**.

While fixing it, I noticed we needed `<fail>` for our own automated test. I'll reinstate `<fail>` in a way the user can disable it.

cirulls commented:
Thanks for your work. Do you want me to have a look at the issue you encountered in https://github.com/xspec/xspec/pull/160/commits/1652d303db416c525f70eb1fdcb86d637915f8cd for Travis? Not sure I'll go any further than you but I can investigate if it helps. 

@AirQuick: Are there any other changes you would like to implement in this pull request or in other issues? My plan is to release a new version of XSpec before [XML Prague](xmlprague.cz) so this would be the last pull request to merged before the new release. Is this fine with you? I know you raised other issues/PR but I wanted to prioritize the Schematron support in this release and unfortunately I didn't have time to work on other things. Since you contributed a lot to this release, I would like to get your go-ahead as well.

AirQuick commented:
The issue encountered on Travis was not reproduced locally even with the same version (1.9.3) of Ant. Something may be special on Travis's Ant or environment, I don't know. So I had to install a cleaner Ant on Travis.

While doing so, I noticed Bats was readily available on Travis. So I removed its installation process from `.travis.yml`.

As we've put some installation processes, `.travis.yml` and `appveyor.yml` are getting crowded. So I tidied them up a bit.

`<fail>` is reinstated in 52c123c1b5110815de46964a271c875a57eb7ab6.
With `<fail>`, I added a few more tests in `xspec.bats` and `xspec-bat.cmd`. In the course of adding the tests, I had to inject a test failure into `tutorial/schematron/demo-04.xspec`.

I hope I'm done. Please review the changes which have been accumulated unexpectedly. Beware; I'm not familiar with Ant :)

> this would be the last pull request to be merged before the new release. Is this fine with you?

No problem.

AirQuick commented:
* Fixed a bug that `xspec.dir` is not always honored in Schematron.
* Removed yet another unnecessary parameter `xspec.compiled.xsl`.
* I believe `filePrefix` comments in https://github.com/xspec/xspec/pull/138 are obsolete.

I hope I'm really done now :(

Various changes have been made since the initial commit. For writing the usage in Wiki, you might want to take a look at `xspec.sh`, `xspec.bat` and an AppVeyor build log (actual command lines are printed there) as well as the usage comment in `build.xml`.

cirulls commented:
Thanks for your work and the many improvements. I'll review your changes this week and let you know. If everything is fine, I'll approve it and merge this branch into `master`.

AirQuick commented:
Sorry for adding commits. They are supposed to just augment tests and simplify `build.xml`.

cirulls commented:
@AirQuick: Thanks a lot for your work, it's amazing how you improved this original pull request! Thanks for adding additional tests (always appreciated!), for removing unncessary dependencies, and simplifying the build script. Thanks also for making the Travis and AppVeyor configurations closer and for standardizing on tools like `curl`. 

I checked your commits and they look good to me. I just have few questions:

1. I saw you installed Ant 1.9.8 as it is the Ant version installed in oXygen. I read your comment about the trouble you had with Ant installed by default on Travis. For documentation purposes, I'd like to know if we should state that XSpec has been tested with v1.9.8 or if XSpec actually **requires** Ant v1.9.8. I think the former but it would be good if you can confirm.
2. I'll update the documentation on Ant and will ask you to review it.
3. I noticed you introduced some JavaScript in [make-path.js](https://github.com/xspec/xspec/blob/review/138/src/ant/make-path.js) which is called in the [build file](https://github.com/xspec/xspec/blob/review/138/build.xml#L122). Was it problematic to do this with `<makeurl>`? I'm just a bit concerned about introducing another programming language in the code base for maintainability reasons and wonder if the same result could be achieved differently. This shouldn't stop this pull request being merged (which is my priority) but I just wanted to mention it.

Let me know what you think. I'm aiming to merge this pull request during the weekend and then working on #139 for releasing XSpec v1.0.0.

AirQuick commented:
Based on [the Ant manual](https://ant.apache.org/manual/ifunless.html), the new `build.xml` requires Ant 1.9.1. I tested with Ant 1.9.1 and it worked. You can confirm it [here](https://travis-ci.org/AirQuick/xspec/jobs/336434091#L451-L452). So only 1.9.1 or later should be required.
I still do not know why the preinstalled Ant (1.9.3) on Travis fails. The error message is weird. If I install the same version of Ant separately on Travis, [it works](https://travis-ci.org/AirQuick/xspec/jobs/336438529#L451-L452).

`<makeurl>` converts file path to URL. What we need is a reverse function (converts URL to file path). I couldn't find a smarter way of achieving it without external language.

cirulls commented:
Thanks for your replies. I'll mention in the documentation that Ant 1.9.1 is the minimum requirement (I didn't know about `if` and `unless` in Ant, I'm glad they have been introduced so we don't have to use the `ant-contrib` library).

I can see the rationale for using a scripting programming language and I can see it is also one of the suggested solutions for a [similar question on StackOverflow](https://stackoverflow.com/questions/26539422/decode-url-in-ant). It makes sense. 

I'll merge this pull request during the weekend and review the documentation on the wiki.

cirulls commented:
@AirQuick: I merged this pull request into `master` and deleted branch `review/138` (if you think of other changes, please use a new branch and open a new pull request). I updated the [documentation on the wiki](https://github.com/xspec/xspec/wiki/Running-with-Ant), in particular the section [Running XSpec's Ant build file from the command line](https://github.com/xspec/xspec/wiki/Running-with-Ant#running-xspecs-ant-build-file-from-the-command-line). Could you please have a look at some point? Feel free to add changes in the documentation as we can't open pull request there.

AirQuick commented:
Updated the [wiki](https://github.com/xspec/xspec/wiki/Running-with-Ant/_history):
* Removed the section **Using the xspec macro**, because the macro has been [removed](https://github.com/xspec/xspec/pull/160/commits/4e2ed51f98b7f530b21bbbc3742eb97fe9712294#diff-2cccd7bf48b7a9cc113ff564acd802a8L59).
* Moved `lib` from the section **Useful Properties** to its own section, because `lib` is not a property.
* Removed the phrase `and are passed to the Ant build script as option using the syntax -Dproperty-value unless otherwise specified`, because it is included in the list of `You can override the defaults by predefining the property:`.
* Other minor fixes.

markdunnoup commented:
Thanks very much to both @AirQuick and @cirulls for your efforts in improving and locking down this request!

cirulls commented:
@AirQuick: thanks for updating the wiki, it's much clearer now. 
@markdunnoup: I'll close #138.
